overlord/hookstate/handler.go

```diff
@@ -0,0 +1,113 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hookstate
+
+import (
+	"fmt"
+	"sync"
+)
+
+// Handler is the interface a client must satify to handle hooks.
+type Handler interface {
+	// Before is called right before the hook is to be run.
+	Before() error
+
+	// Done is called right after the hook has finished successfully.
+	Done() error
+
+	// Error is called if the hook encounters an error while running.
+	Error(err error) error
+
+	// Get the data that this handler has associated with the given key.
+	Get(key string) (map[string]interface{}, error)
```

Comment:
This is being called as

_HookManager.GetHookData => HandlerCollection.GetHandlerData => Handler.Get_

But it's not clear to me where the data comes from, or where it goes to. It's also not entirely clear how that improves the situation vs. having simply `Context.Get` alone, which is effectively the only place we have to store the data when we don't have an understanding of what the handler or the data actually are, which is the situation for someone running, say, `HookManager.SetHookData`.

I suggest taking a step back and scaling down on the abstractions introduced a bit, and perhaps trying to reach those same ideas via more concrete needs. I believe we don't even need a HandlerCollection for example, as that's pretty much the only role of the HookManager (it starts handlers and tracks them).

What we've got today is this:

  * _Context_ - This is the closest we have to an "instance" of a hook execution. It's dumb, holds some persistent data unaware of its meaning, and general actions the handler may take.
  * _Handler_ - The code that knows what the purpose of the executing hook is. We have one of these for each context. May hold some transient data, but must store persistent run details in the context.
  * _HookManager_ - The thing that creates context and handler, and should track them so we can find them again, and clean up when necessary.

That's looking good so far. We're lacking just one last idea from our master plan: the hook tools. These are the things the hooks can call and that will talk back to snapd for doing all sorts of actions, and when such a tool (command) is run we must somehow manage to have at hand the three involved parties to work with: the tool which is doing the particular action, the handler instance which is managing the hook execution, and the context they're running for. Starting with that, we can then tune and implement whatever logic we need on the several use cases we know about.

Here are a few very simple steps to concretely bootstrap that last phase:

  1. Let's introduce a trivial snaptool command.
  1. When `snaptool foo --bar` is called, we POST `{"context": $SNAP_CONTEXT", "args": ["snaptool", "foo", "--bar"]}` to `/v2/snaptool`
  1. $SNAP_CONTEXT is internally available as Context.Key, and is built as a random 32 char hex string.
  1. The data is unmarshaled into &hookstate.ToolRequest, and handed to the manager as HookManager.RunTool(request)
  1. The HookManager then parses the arguments in the request as usual for commands (similar to cmd/snap/main.go but hopefully simpler) and calls something along the lines of toolCmd.Run(handler, context, args)
  1. The tool needs to have a couple of buffers representing stdout and stderr at hand, and their content is sent down as `{"stdout": ..., "stderr": ...}` at the end of the run.

This closes the loop. We then have the tool, the handler, and the context all in the same function, and can implement whatever logic is necessary.